pennylane_lightning_gpu/src/util/MPIManager.hpp

```diff
@@ -0,0 +1,551 @@
+        isExternalComm_ = true;
+        PL_MPI_IS_SUCCESS(MPI_Comm_rank(communiator_, &rank_));
+        PL_MPI_IS_SUCCESS(MPI_Comm_size(communiator_, &size_));
+        findVendor();
+        findVersion();
+        getNumProcsPerNode();
+    }
+
+    MPIManager(MPI_Comm communicator) : communiator_(communicator) {
+        isExternalComm_ = true;
+        PL_MPI_IS_SUCCESS(MPI_Comm_rank(communiator_, &rank_));
+        PL_MPI_IS_SUCCESS(MPI_Comm_size(communiator_, &size_));
+        findVendor();
+        findVersion();
+        getNumProcsPerNode();
+    }
+
+    MPIManager(int argc, char **argv) {
+        PL_MPI_IS_SUCCESS(MPI_Init(&argc, &argv));
+        isExternalComm_ = false;
+        communiator_ = MPI_COMM_WORLD;
+        PL_MPI_IS_SUCCESS(MPI_Comm_rank(communiator_, &rank_));
+        PL_MPI_IS_SUCCESS(MPI_Comm_size(communiator_, &size_));
+        findVendor();
+        findVersion();
+        getNumProcsPerNode();
+        check_mpi_config();
+    }
+
+    MPIManager(const MPIManager &other) {
+        isExternalComm_ = true;
+        rank_ = other.rank_;
+        size_ = other.size_;
+        communiator_ = other.communiator_;
+        vendor_ = other.vendor_;
+        version_ = other.version_;
+        subversion_ = other.subversion_;
+        size_per_node_ = other.size_per_node_;
+    }
+
+    ~MPIManager() {
+        if (!isExternalComm_) {
+            int initflag;
+            int finflag;
+            PL_MPI_IS_SUCCESS(MPI_Initialized(&initflag));
+            PL_MPI_IS_SUCCESS(MPI_Finalized(&finflag));
+            if (initflag && !finflag) {
+                PL_MPI_IS_SUCCESS(MPI_Finalize());
+            }
+        }
+    }
+
+    // General MPI operations
+    /**
+     * @brief Get the process rank in the communicator.
+     */
+    auto getRank() const { return rank_; }
+
+    /**
+     * @brief Get the process number in the communicator.
+     */
+    auto getSize() const { return size_; }
+
+    /**
+     * @brief Get the number of processes per node in the communicator.
+     */
+    auto getSizeNode() const { return size_per_node_; }
+
+    /**
+     * @brief Get the communicator.
+     */
+    auto getComm() const { return communiator_; }
+
+    /**
+     * @brief Get an elapsed time.
+     */
+    auto getTime() const { return MPI_Wtime(); }
+
+    /**
+     * @brief Get the MPI vendor.
+     */
+    auto getVendor() const { return vendor_; }
+
+    /**
+     * @brief Get the MPI version.
+     */
+    auto getVersion() -> std::tuple<int, int> {
+        return {version_, subversion_};
+    }
+
+    /**
+     * @brief Find the MPI vendor.
+     */
+    void findVendor() {
+        char version[MPI_MAX_LIBRARY_VERSION_STRING];
+        int resultlen;
+
+        PL_MPI_IS_SUCCESS(MPI_Get_library_version(version, &resultlen));
+
+        if (strstr(version, "Open MPI") != nullptr) {
+            vendor_ = "Open MPI";
+        } else if (strstr(version, "MPICH") != nullptr) {
+            vendor_ = "MPICH";
```

Comment:
We should use the CPP calls for these rather than the C calls, if the plan is to use them as is. If another C++ only option exists that is faster, we should use that instead.